---
datetime: "2017-02-13T17:00Z"
title: "Promises: Chains"
---
In the previous blog post I told you that the promise which fetch retuns becomes
resolved when reponse headers have been received, before the body has been
received. This gives you a chance to inspect the status and the type of
the response to decide what to do with the rest of it. It's an important
optimisation for large responses which can be processed in chunks like video or
audio. Without this step in between, the entire response would need to be
received before processing of it can be started.

```javascript
function onResponse(res) {
  if (!res.ok) {
    return console.error(`Unexpected status code: ${res.status}`);
  }

  res.text()
    .then(function (text) {
      console.log(text);
    });
}

function onError(error) {
  console.error(`Got error: ${error.stack}`);
}

fetch('https://qubyte.codes')
  .then(onResponse, onError);
```

In this example I use `res.ok`, which evaluates to true when `res.status` is
between `200` and `299` (inclusive). If that's not the case then an error is
logged. When the server responds with a success, the `text` method of the
response is called. This also returns a promise, which resolves once the
response has finished and the response body has been parsed as text.

This nesting of promises isn't great. It doesn't take a lot of imagination to
see this becoming a _callback hell_.

It can be improved:

```javascript
function onResponse(res) {
    if (!res.ok) {
      throw new Error(`Unexpected status code: ${res.status}`);
    }

    return res.text();
}

function onText(text) {
  console.log(text);
}

function onError(error) {
  console.error(`Got error: ${error.stack}`);
}

fetch('https://qubyte.codes')
  .then(onResponse)
  .then(onText, onError);
```

This example exploits a clever feature of the `then` method. `then` returns a
new promise! The behaviour of this promise can appear complex so I'll try to
describe it as precisely as possible. Don't worry, it'll become second nature.

Given a promise `o`, when `o.then` is called with a resolution and a rejection
handler, a promise `p` is returned by `o.then` with the following behaviour:

 - When either handler is called and returns a promise `q`, `p` will resolve
   when `q` resolves, or it will reject when `q` rejects. The resolution or
   rejection value of `p` is taken from `q`.
 - When either handler returns a value, `p` will resolve with that value.
 - When either handler throws an error, `p` will reject with that error.
 - When there is no resolution handler, `p` will resolve when the original
   promise `o` resolves, with the same value (passthrough).
 - When there is no rejection handler, `p` will reject when the original promise
   `o` rejects, with the same error value (passthrough).

These behaviours allow you to handle resolutions and rejections in a stepwise
fashion. This behaviour is important for this part:

```javascript
fetch('https://qubyte.codes')
  .then(onResponse)
  .then(onText, onError);
```

In the best case scenario, the fetch resolves with a successful response object
and `onResponse` is called with it. That function returns a promise which
resolves when the response body has been downloaded and parsed. With the
response body parsed, `onText` is called with the resultant text and the text is
logged.

In the worst case scenario, fetch rejects (perhaps there was a communication
error of some kind). Since the first `then` has no rejection handler, the error
passes through for `onError` to handle, and the error is logged.

In the middle case, the request resolves, but with an unsuccessful status code.
`onResponse` throws an error, which `onError` receives and logs.

In the list above I distingish between values and promises, but this is not the
whole story. The distinction can be seen with the behaviour of
`Promise.resolve`. `Promise` itself will be discussed in the next post in this
series. For this post it is sufficient to know that all native promises are
instances of `Promise`. i.e.

```javascript
fetch('https://qubyte.codes') instanceof Promise; // evaluates to true
```

Since there are many implementations of promises which predate native promises,
they need to be able to work together. `Promise.resolve` is a mechanism which
coerces anything to a promise.

Simple values like numbers, booleans, string, arrays, etc. are wrapped in a
resolved promise.

```javascript
var p = Promise.resolve('Hello, world!');

p.then(val => console.log(val)); // This will log "Hello, world!".
```

Native promises are returned.

```javascript
var p = fetch('https://qubyte.codes');
var q = Promise.resolve(p);

if (p === q) {
  console.log('Promise.resolve returns native promises.');
}
```

Promise-like objects, often called _thenables_ are coerced. In this case
`Promise.resolve` makes a new promise, and registers resolution and rejection
handlers with the thenable. Then the resolution handler is called, the new
promise resolves with the resolution value. When the rejection handler is
called, the new promise rejects with the rejection value.
